internal/stores/hostdb.go

```diff
@@ -181,3 +195,236 @@ func NewJSONHostDB(dir string) (*JSONHostDB, modules.ConsensusChangeID, error) {
 	}
 	return db, ccid, nil
 }
+
+// consensusInfoID defines the primary key of the entry in the consensusInfo
+// table.
+const consensusInfoID = 1
+
+type (
+	// SQLHostDB is a helper type for interacting with an SQL-based HostDB.
+	SQLHostDB struct {
+		staticDB *gorm.DB
+	}
+
+	// host defines a hostdb.Interaction as persisted in the DB.
+	host struct {
+		PublicKey     []byte         `gorm:"primaryKey"`
+		Announcements []announcement `gorm:"foreignKey:Host;references:PublicKey"`
+		Interactions  []interaction  `gorm:"foreignKey:Host;references:PublicKey"`
+	}
+
+	announcement struct {
+		ID          uint64 `gorm:"primaryKey"`
+		Host        []byte
+		BlockHeight uint64
+		BlockID     []byte
+		Timestamp   time.Time
+		NetAddress  string
+	}
+
+	// interaction defines a hostdb.Interaction as persisted in the DB.
+	interaction struct {
+		ID     uint64 `gorm:"primaryKey"`
+		Type   string
+		Result json.RawMessage
+
+		// Host and Timestamp form a composite index where Hosts has the
+		// higher priority.
+		Host      []byte    `gorm:"index:idx_host:2"`
+		Timestamp time.Time `gorm:"index:idx_host:1"`
+	}
+
+	// consensusInfo defines table which stores the latest consensus info
+	// known to the hostdb. It should only ever contain a single entry with
+	// the consensusInfoID primary key.
+	consensusInfo struct {
+		ID   uint8 `gorm:"primaryKey"`
+		CCID []byte
+	}
+)
+
+// Host converts a host into a hostdb.Host.
+func (h host) Host() hostdb.Host {
+	hdbHost := hostdb.Host{}
+	copy(hdbHost.PublicKey[:], h.PublicKey)
+	for _, a := range h.Announcements {
+		hdbHost.Announcements = append(hdbHost.Announcements, a.Announcement())
```

Comment:
`nit` avoid appends - might be worthwhile to do this in the case of transformer functions

Reply:
How would you avoid that?

Reply:
I think PJ means preallocating the slice:
```go
hdbHost := hostdb.Host{
    Announcements: make([]hostdb.Announcement, len(h.Announcements)),
    Interactions:    make([]hostdb.Interaction, len(h.Interactions)),
}
```